### ISSUE TITLE
Bug in ExclusiveSum

### ISSUE THREAD

**Issue by csukuangfj** [NVIDIA/cub#302]
The following code prints
```
0 100 -128 0
```

NOTE: It overflows!!!!

```
nvcc --version
```
prints

```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2021 NVIDIA Corporation
Built on Sun_Mar_21_19:15:46_PDT_2021
Cuda compilation tools, release 11.3, V11.3.58
Build cuda_11.3.r11.3/compiler.29745058_0
```

```cpp
#include <vector>

#include "cub/cub.cuh"

int main() {
  std::vector<int8_t> v = {100, 28, 1};

  int8_t *src;
  // ignore error check
  cudaMalloc((void **)&src, 3);
  cudaMemcpy(src, v.data(), 3, cudaMemcpyHostToDevice);

  int32_t *dst;
  cudaMalloc((void **)&dst, 4 * sizeof(int32_t));

  void *d_temp_storage = nullptr;
  size_t temp_storage_bytes = 0;
  cub::DeviceScan::ExclusiveSum(d_temp_storage, temp_storage_bytes, src, dst,
                                3);
  cudaMalloc(&d_temp_storage, temp_storage_bytes);
  cub::DeviceScan::ExclusiveSum(d_temp_storage, temp_storage_bytes, src, dst,
                                3);

  int32_t h[4];
  cudaMemcpy(h, dst, 4 * sizeof(int32_t), cudaMemcpyDeviceToHost);

  for (int32_t i = 0; i != 4; ++i) {
    std::cout << h[i] << " ";
  }

  std::cout << "\n";
  // ignore memory leak
  return 0;
}
```

**Comment by elstehle** [comment]
@csukuangfj, you can use a `TransformInputIterator` to make sure that your input elements are correctly casted to a type that won't overflow. As a result, the algorithm will never "see" the original `int8_t` and therefore is not inclined to use that type for the "accumulator".

E.g.,
```cpp
#include <vector>

#include "cub/cub.cuh"

template <typename TargetT>
struct TypeConversionOp {
    template <typename InputT>
    __host__ __device__
    TargetT operator()(const InputT& v) const
    {
        return static_cast<TargetT>(v);
    }
};

int main()
{
    std::vector<int8_t> v = { 100, 28, 1, 3 };

    int8_t* src;
    // ignore error check
    cudaMalloc((void**)&src, v.size());
    cudaMemcpy(src, v.data(), v.size(), cudaMemcpyHostToDevice);

    int32_t* dst;
    cudaMalloc((void**)&dst, v.size() * sizeof(int32_t));

    void* d_temp_storage = nullptr;
    size_t temp_storage_bytes = 0;
    cub::TransformInputIterator<int32_t, TypeConversionOp<int32_t>, int8_t*> in_it(src, TypeConversionOp<int32_t>());
    cub::DeviceScan::ExclusiveSum(d_temp_storage, temp_storage_bytes, in_it, dst,
        v.size());
    cudaMalloc(&d_temp_storage, temp_storage_bytes);
    cub::DeviceScan::ExclusiveSum(d_temp_storage, temp_storage_bytes, in_it, dst,
        v.size());

    int32_t h[v.size()];
    cudaMemcpy(h, dst, v.size() * sizeof(int32_t), cudaMemcpyDeviceToHost);

    for (int32_t i = 0; i != v.size(); ++i) {
        std::cout << h[i] << " ";
    }

    std::cout << "\n";
    // ignore memory leak
    return 0;
}
```

**Comment by allisonvacanti** [comment]
@elstehle is right -- this is expected behavior, and the transform iterator is an easy way to get the behavior you're looking for. Using mixed types in scans requires some caution, since the intermediate type used to accumulate the scan result may not be what you expect. As you've noticed, this can easily lead to overflow and other numeric stability issues.

Another workaround available in CUB 1.11.0 and CTK 11.3 is to use the `ExclusiveScan` overload with an explicit initial value and the `cub::Sum` functor. That version of CUB updated the scan algorithms to follow https://wg21.link/P0571, which defines how the intermediate accumulator type is chosen.

The rules are:

1) If an `initial_value` is specified, the `initial_value`'s type is used to accumulate intermediate results.
2) If using an overload that doesn't provide an `initial_value`, the input value type is used.

tl;dr this can be fixed by either (a) transforming the input to a larger type (as @elstehle suggested), or (b) specify the desired accumulator type explicitly via the `initial_value` parameter of `ExclusiveScan`.

**Comment by csukuangfj** [comment]
@elstehle @allisonvacanti 

Thank you for proposing the possible solutions.

However, I still consider it is a bug or a serious flaw in your API design.

The reasons are given below:

1. My above test code works perfectly for CUDA 10.1, 10.2, and 11.0. It breaks only in CUDA 11.2 and 11.3.

   As far as I can tell, CUB is not included in the cudatoolkit 10.1 and 10.2 and is included in cudatoolkit 11.x.
   So there must be some backward-incompatible changes between CUDA 11.0 and CUDA 11.2 since the 
   code works for CUDA 11.0, but not for CUDA 11.2.  The change may break existing code in various projects
   that are using CUB.

2. Nothing is said in https://nvlabs.github.io/cub/structcub_1_1_device_scan.html#a02b2d2e98f89f80813460f6a6ea1692b
    that there may be possible overflow in `ExclusiveSum`.

     Only you two, and other core developers of CUB know this issue since it is an implementation detail.
     You are providing an error-prone API for end-users. Furthermore, it has different behaviors in different
     cudatoolkit versions and it is not documented explicitly.

3. The above test code is just a demo to show this issue. Our actual usage is templated on the arguments.

    The provided API forces users to handle possible overflow by making their code complicated. When not handled
    appropriately, it may cause runtime overhead for unnecessary conversions.

**Comment by jrhemstad** [comment]
@csukuangfj there is nothing unique to CUB about this behavior. `std::exclusive_scan` demonstrates the same exact behavior. 

https://godbolt.org/z/7G38nhP4T

It is not canonical C++ to guard against overflow for the caller. It is the caller's responsibility to recognize a situation that may lead to overflow and take appropriate action (such as casting to a wider type).

**Comment by csukuangfj** [comment]
@jrhemstad 

Is there a reason that ExclusiveSum shows different behaviors using CUB from CUDA 11.0 and 11.3?

I don't think `std::exclusive_scan` behaves differently among different C++ standards.

**Comment by jrhemstad** [comment]
Integer overflow is undefined behavior, which means an implementation is free to do whatever it wants in those situations and behavior can change at any time.

**Comment by allisonvacanti** [comment]
> Is there a reason that ExclusiveSum shows different behaviors using CUB from CUDA 11.0 and 11.3?

Yes; the implementation changed in 11.3 as detailed [in an earlier comment](https://github.com/NVIDIA/cub/issues/302#issuecomment-838765678).

The CUB implementation's behavior for mixed-type scans was never well defined previously, and the new implementation follows guidance from the C++ standards committee for providing well-defined behavior in these edge cases.

Pre-11.3, a different type was deduced for the accumulator, but it was no less likely to overflow. Whether or not the accumulator will overflow depends on:

1. The static type of the input.
2. The number of elements in the input set.
3. The values in the input set.

CUB must determine the accumulator type during compilation, and given the amount of runtime variability in the last two points, we cannot guarantee that any particular accumulator type will not overflow. These are usecase-specific details that only a CUB user would reasonably know.

So the choice of accumulator type set to a sensible default using the approach defined in https://wg21.link/P0571, which also provides consistent mechanisms for users to influence the deduced accumulator, since they know their usecases far better than we do.

**Comment by csukuangfj** [comment]
@allisonvacanti @jrhemstad 

Thank you for the explanation and the link to the standard document.
